app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/ExtractDataFromCategoryTree.php

```diff
@@ -42,14 +49,61 @@ public function execute(\Iterator $iterator): array
             /** @var CategoryInterface $category */
             $category = $iterator->current();
             $iterator->next();
-            $nextCategory = $iterator->current();
-            $tree[$category->getId()] = $this->categoryHydrator->hydrateCategory($category);
-            $tree[$category->getId()]['model'] = $category;
-            if ($nextCategory && (int) $nextCategory->getLevel() !== (int) $category->getLevel()) {
-                $tree[$category->getId()]['children'] = $this->execute($iterator);
+
+            $pathElements = explode("/", $category->getPath());
+            $this->iteratingCategory = $category;
+
+            $currentLevelTree = $this->generateLevelTree($pathElements, self::START_CATEGORY_FETCH_LEVEL);
+            if (empty($tree)) {
+                $tree = $currentLevelTree;
             }
+            $tree = $this->mergeCategoriesTrees($currentLevelTree, $tree);
         }
 
         return $tree;
     }
+
+    /**
+     * Merge together complex categories tree
+     *
+     * @param array $tree1
+     * @param array $tree2
+     * @return array
+     */
+    private function mergeCategoriesTrees(array &$tree1, array &$tree2): array
+    {
+        $mergedTree = $tree1;
+        foreach ($tree2 as $currentKey => &$value) {
+            if (is_array($value) && isset($mergedTree[$currentKey]) && is_array($mergedTree[$currentKey])) {
+                $mergedTree[$currentKey] = $this->mergeCategoriesTrees($mergedTree[$currentKey], $value);
+            } else {
+                $mergedTree[$currentKey] = $value;
+            }
+        }
+        return $mergedTree;
+    }
+
+    /**
+     * Recursive method to generate tree for one category path
+     *
+     * @param $elements
+     * @param $index
+     * @return array
+     */
+    private function generateLevelTree($elements, $index): array
```

Comment:
Method name is not reflecting its purpose, need to read the docs to understand what it is doing. Please consider improving method name.